### DISCUSSION TITLE
Introduce Steps based separation for big test methods

### DISCUSSION THREAD

**Discussion by nithinbandaru1** [nunit/nunit#4869] [Ideas]
Background:

1. When we write tests at higher level like Integration Tests or User Acceptance Tests, basically the are complicated tests which will have different steps to successfully run the test.
2. With this type of tests there will be lot of common code that need to be executed before the actual scenario can be tested.
3. To reduce or avoid duplicate code currently available techniques are beforeTest method or common helper classes with methods etc.

Problem Statement:

1. Now if the test fails it is not very straight forward to find out where is the issue at which step the issue occurred.
2. Different platforms like Jira, Azure DevOps supports steps level tracking of tests and the results cannot be captured and pushed to tests at this level now.
3. This is basically forcing people to use different frameworks like Specflow which asks us to write things in Gherkin first and then write glue code which is not simple as writing basic NUnit tests and for publish step level test results we need to use different tool.

Solution Proposal:

1. Introduce new attributes like TestStep or a bit further like Given, When, Then inheriting TestStep if C# attributes supports inheriting.
2. I think introducing new attributes is quite straight forward but there will be need for change in test runner provided by nunit.
3. How complex it is to achieve this kind of test code processing. Is it practical to figure out if methods called in method marked with [Test] and capture results for each method call separately by the test runner.  Or maybe introducing some builder pattern for test would be straight forward but I guess that is what SpecFlow does to a certain extent with generated code atleast here it will be bit straight forward because of no Gerkin.
4. How flexible it is to modify the runner to add this step level execution ability? If this feature is there then NUnit will become a goto framework for all level of tests.

**Comment by OsirisTerje** [comment]
You can actually do this using the current available attributes.  And you can enhance it by making some small changes yourselves.

If you regard one TestFixture as one integrationtest, and then each Test as a Step:

The code below will execute the tests in the order you require, and by simply naming them appropriately you achieve what you want functionally:

```cs
public class MyIntegrationTest
{
    [Test,Order(1)]
    public void Step1()
    {
        TestContext.Out.WriteLine(nameof(MyIntegrationTest)+"."+nameof(Step1));
        Assert.Pass();
    }

    [Test, Order(2)]
    public void Step2()
    {
        TestContext.Out.WriteLine(nameof(MyIntegrationTest) + "." + nameof(Step2));
        Assert.Pass();
    }

    [Test, Order(3)]
    public void Step3()
    {
        TestContext.Out.WriteLine(nameof(MyIntegrationTest) + "." + nameof(Step3));
        Assert.Pass();
    }
}
```

But you can go further!  Create your own Step attribute, and let that handle both the Test and the Order attribute, sort of hiding them so that the test actually look more like what you describe above.

First the Step:

```cs
[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
public class StepAttribute(int order) : TestAttribute
{
    public new void ApplyToTest(NUnit.Framework.Internal.Test test)
    {
        base.ApplyToTest(test);

        // Apply the order to the test
        test.Properties.Set(NUnit.Framework.Internal.PropertyNames.Order, order);
    }
}
```

And then use this in the integrationtest, I have added an extra int there so that it is obvious when you run it that it actually executes in correct order:

```cs
public class MyIntegrationTest2
{
    int i = 0;
    [Step(1)]
    public void Step1()
    {
        TestContext.Out.WriteLine($"{nameof(MyIntegrationTest2)}.{nameof(Step1)} {++i}");
        Assert.Pass();
    }

    [Step(3)]
    public void Step3()
    {
        TestContext.Out.WriteLine($"{nameof(MyIntegrationTest2)}.{nameof(Step3)} {++i}");
        Assert.Pass();
    }

    [Step(2)]
    public void Step2()
    {
        TestContext.Out.WriteLine($"{nameof(MyIntegrationTest2)}.{nameof(Step2)} {++i}");
        Assert.Pass();
    }
}
```

Running these two using `dotnet test`, clearly displays the names and order:
![image](https://github.com/user-attachments/assets/42e786d9-c769-44eb-b1d1-62008916fbc5)

The code above is also here as a working project:  https://github.com/nunit/nunit.issues/tree/main/Discussion4869

You can then of course extend further with Given, When, Then, and if you like make your own TestFixture attribute, like calling it IntegrationTest.
